File: Data Structures & Algorithms/climbing-stairs/submission-1.py
class Solution:
    def climbStairs(self, n: int) -> int:
        if n <= 2:
            return n

        def dfs(i):
            if i >= n:
                return i == n

            return dfs(i + 1) + dfs(i + 2)

        return dfs(0)

File: Data Structures & Algorithms/house-robber/submission-1.cpp
class Solution {
    std::unordered_map<int, int> dfs_values;

   public:
    int rob(vector<int>& nums) {
        size_t n = nums.size();
        dfs_values.reserve(n);

        if (n == 1) {
            return nums[0];
        }
        
        return dfs(nums, 0);
    }

    int dfs(std::vector<int>& nums, int i){
        if (i >= nums.size()) {
            return 0;
        }

        if (const auto& it = dfs_values.find(i); it != std::end(dfs_values)) {
            return it->second;
        } else {
            const int res = std::max(dfs(nums, i + 1), nums[i] + dfs(nums, i + 2));
            dfs_values.emplace(i, res);
            return res;
        }
    };
};
